### ISSUE TITLE
Kohana Retirement – 2017/07/01

### ISSUE THREAD

**Issue by lermontex** [kohana/kohana#110]
@enov, @rjd22, @acoulton, @biakaveron, @BRMatt, @cbandy, @dm, @isaiahdw, @jheathco , @kemo, @lenton, @nodren, @randomecho, @Zeelot, @zombor, @JackEllis, @melvinmt, @sergeyklay, @sarciszewski, ...

http://discourse.kohanaframework.org/t/kohana-retirement-2017-07-01/1277

Many of you use Kohana in your projects, is it really lost?

I found one active repository that created @neo22s : https://github.com/koseven
Perhaps you can help? There are other analogues?

In general, the repository has a branch with support for namespases (https://github.com/kohana/core/tree/feature/apply-namespaces),  @rjd22  You do not have the motivation to continue working?

Are you interested in continuing with the framework? What are your suggestions?

If there are no analogues, and Kohana retires, welcome to [Koseven](https://github.com/koseven).

I think together we can keep the framework up to date, even under a different name.

Thank you!

**Comment by JackEllis** [comment]
I do the majority / all of my PHP work with Laravel now. And to be honest, the seniors in the Kohana community moved to clean code where the delivery mechanism was irrelevant and the separation of controllers, business rules, frameworks etc. is more important.

I'll still work on legacy Kohana projects if I'm asked but I certainly wouldn't start a new project in Kohana nowadays.

**Comment by neo22s** [comment]
Hello,

@lermontex thanks for the mention. I do as well moved to laravel for new apps as @JackEllis comments but I do have 1 big project I can not afford to move to laravel at this time, nor in a near future. Thats why Koseven.

We are already few and soon to be released a new version with many improvements and bug fixes, should be like a drop in replacement with API compatible with KO except the exception system which was changed to make it work on PHP7.

So far really happy how is it going wish I had more time to spend on this.

Please join us at our telegram channel https://telegram.me/koseven or feel free to open any issue.

Thanks a lot!

**Comment by rjd22** [comment]
@lermontex -  Like @JackEllis and @neo22s I've moved to Symfony framework for my work. Kohana is too far behind on the other frameworks so you're better off migrating to Symfony or Laravel for your next project.

Those communities are more alive and there is really good support for integration with other systems. They also follow the same principles that Kohana Framework did. You can develop application while leveraging the framework but you don't have to build your applications on top of the framework.

I still love Kohana for what it did for me. Working with it was a great experience, but it's not good enough anymore for where the community stands right now.

**Comment by sigma-technology** [comment]
@lermontex @JackEllis @neo22s  I have seen this coming for a while now, although I was secretly hoping it would never happen.

I found Kohana incredible mainly because of the **Hierarchical** MVC, which was well done IMO.  I'm looking into Symfony and Laravel as replacements, but will use Koseven (thank you @neo22s) in the interim.  A couple of questions, if you'd be so kind:

1. Do either of these frameworks have the same/better level of HMVC functonality?
2. Which one is better supported? - having migrated to Kohana from Code Igniter (after it died) whichever one we switch to, I want to avoid this sort of thing happening again for a long time.

I am aware that it's going to be very difficult for me to find web developers to join the team that have Kohana experience, and given the lack of good documentation for Kohana, it's going to be a constant uphill battle.  I would really appreciate any advice you can give me based on your experiences.

**Comment by neo22s** [comment]
@sigmatec laravel community and tools they have built around are just amazing... plus the docs video courses etc... ;)

**Comment by nodren** [comment]
@sigmatec Symfony supports HMVC much in the same way Kohana does, there's a render tag in twig that allows you to call a controller action.

I'm unaware if laravel also has documented support for it, but it uses symfony components under the hood, and I imagine if you did a dive into either documentation or code(or even a google search) you could find a way of doing this in laravel as well.

**Comment by rjd22** [comment]
@nodren The difference between Laravel en Symfony is Convention over Configuration. 

Laravel - mostly focuses on working out of the box but when you do so you loose some flexibility.
Symfony - mostly focuses on being flexible from the start but it means you have to configure some stuff to get it working.

The learning curve of Symfony is also higher than of Laravel.

**Comment by acoulton** [comment]
tl;dr: dropping the github repos will cause people pain and hasn't been publicised anywhere like enough.

@shadowhand I was pretty disappointed to see this, I think the discourse post falls far short of an appropriate level of announcement that the github organisation is going to be closed. If it wasn't for @lermontex I'd not have any idea this was planned, and I'm a contributor.

Fair enough deleting the website, forums, etc - I get there's cost and hassle involved in hosting them. I can't really see any justification though for deleting the (free) github organisation and repos.

When you do that, the **download URLs published by packagist will also break**. Anyone who's using the packages via composer will continue to be able to run `composer install` on an existing machine indefinitely without any warning or error. They will only get a failed install when they next try to deploy to a machine with an empty composer cache.

If that's part of a planned update or rotation, that'll be irritating but solvable assuming they have a copy of the code somewhere. Bearing in mind it could be some time past 1st July and this thread and the retirement forum post will have dropped off the internet so working out what happened and where to look will be tricky. I've had personal experience of this when you deleted the shadowhand/email and shadowhand/uuid modules, but they were at least pretty easy to replace.

If that's part of dealing with unexpected downtime, it'll be a massive headache to suddenly find that critical project dependencies are AWOL.

Migrating to koseven or a fork is fine, but isn't something you want to do in a hurry and will obviously then involve the circular dependency hell of updating any other dependencies that depend on kohana/xx because even one reference to a kohana/* package will cause composer install to fail.

I realise arguably this is a weakness in composer/packagist that people should be preparing for anyway, but lots of people don't and the standard documented composer flow that most people use doesn't really consider that eventuality.

As such, I feel quite strongly that when you publish code to packagist there's a fairly significant moral obligation to leave that code there in perpetuity, particularly if it's costing nothing to keep it available.

If you're not willing to do that then I think there needs to be a much more concerted effort to let people know this is coming. For a start, the packagist packages need to be marked as abandoned, but even that won't leap out at users of legacy projects.

**Comment by acoulton** [comment]
Just to clarify, I'm also moving away from Kohana for new projects - to a mix of things, primarily using individual components and a clean-code isolated-dependency setup.

But we are still supporting a number of critical legacy kohana applications that are running just fine - particularly since over the years we've been gradually decoupling them from the framework. We might in due course move them to koseven when I've had a chance to look at that, but short-medium term would really be planning to stick with our ecosystem of kohana/* based packages.

If it helps, I'm more than happy to take over administration of the github organisation.

**Comment by sigma-technology** [comment]
@neo22s @nodren @rjd22 Thank you for your opinions on the two. After some extensive research I've decided to go with Symfony as it seems to be more flexible and rigorously supported. Now we begin the horrible task of migrating the system to be underpinned by Symfony instead of Kohana. I can already see some similarities though so hopefully it won't be too painful.

**Comment by nodren** [comment]
Couple things. 

First, why do you have to migrate from kohana? It's not like there is any major security risks you would be vulnerable to by staying put. 

Secondly, if you too my comment too positively for symfony, I recommend taking a second look at your options.  Let me preface what I'm saying here to say that I don't prefer symfony and would rather use almost anything else instead. It's overly verbose and tries way to hard for an enterprise java esque syntax, also doctrine is not an active record pattern similar to the kohana orm or laravels eloquent. 

These changes make any kind of migration inherently more difficult plus the learning curve and style is completely at odds with kohana. 

If laravel is a choice for you, I highly recommend you look much closer at that choice given your kohana background. 

All that said, switching frameworks is a nuclear option that you should avoid, because your looking at a full rewrite of your project for what would amount to very little gain. Consider using something else, like laravel, for a future project and leave your kohana project alone

**Comment by sigma-technology** [comment]
@nodren We are still actively developing this system, so it just feels like continuing with Kohana is eventually going to lead to issues or road blocks somewhere down the road, at which point it will have to be rewritten anyway.  I also need to recruit additional developers to support the project, and it's going to be very difficult to find Kohana devs.  I agree it's a nuclear option but this piece of software is our entire business so we have to ensure the longevity of it; I can already see advantages in Symfony that will prove useful over Kohana.

However, I had assumed Doctrine was pretty much the same as Kohana's ORM, so there's clearly more research required.  I will investigate Laravel in more detail before making a final decision; it's definitely between the two of them.  Thank you for challenging my assumptions.

**Comment by entermix** [comment]
Why do you think that it's difficult to find developers for Kohana?

And if after a while Symfony also stops support? Will you switch back to another framework?

**Comment by entermix** [comment]
@nodren, 
Why do you think that it's difficult to find developers for Kohana?
And if after a while Symfony also stops support? Will you switch back to another framework?

**Comment by entermix** [comment]
I support @nodren
@sigmatec, Why do you think that it's difficult to find developers for Kohana?

I do not want to persuade you, but if after some time Symfony/Laravel also stops support? Will you switch back to another framework?

**Comment by nodren** [comment]
@sigmatec Here's some links to add your research.

Doctrine uses the [Data Mapper pattern](https://en.wikipedia.org/wiki/Data_mapper_pattern) while Kohana ORM, and Laravel Eloquent use the [Active Record Pattern](https://en.wikipedia.org/wiki/Active_record_pattern). 

Naturally this is entirely up to personal preference, however you'll find the Data mapper pattern is significantly more restrictive compared to Active Record as far as what you can do with it and what type of operations are allowed in what part of your code base.  You'll also notice Active record is the more "popular" approach, which if your concern is recruiting additional developers, you'll find many will be more at home with the Active record approach of Laravel's eloquent compared to the Data mapper approach of doctrine

I also wouldn't recommend dropping doctrine for eloquent but keeping symfony, You're solving one problem but will likely find yourself adding about 40 others over the course of development

**Comment by shadowhand** [comment]
@acoulton my only reason for terminating the Github project is that I no longer want to manage it. If you and @neo22s want to collaborate on that front, I am willing to retire as BDFL and designate the two of you as "archivists" or something along those lines.

@neo22s has already agreed to take ownership of the domain and will be paying for future registration.

To be clear: my involvement with Kohana is going to be finished, and Kohana will no longer be part of the Software Conservancy. This will remove the ability for Kohana to accept any kind of donations  as a non-profit organization and will eliminate all paid hosting. Any and all future hosting will have to be free (Github pages?) or paid for by someone else.

**Comment by sigma-technology** [comment]
@entermix I have already had issues trying to find competent Kohana developers (in the UK).  I am only assuming that Symfony/Laravel developers would be easier to find because they are the most popular frameworks.  If Symfony/Laravel stopped support I would not rule out having to switch again if it ensured the future of the company, but I understand and accept the reasoning behind your question.

@nodren Thank you for the links, an Active Record ORM is definitely going to be the most advantageous for us, and I wouldn't even want to attempt using Eloquent with Symfony.  Looks like I'm going to be spending the rest of the day reading up on, and watching videos about Laravel.

**Comment by acoulton** [comment]
@shadowhand thank you. Understand you want to stand down entirely - and that's fine, thank you for all your work over the years. I'm happy to take responsibility for the github org (with/without @neo22s as required).

**Comment by acoulton** [comment]
@sigmatec IMO it's difficult to find competent PHP devs at all in the UK, not just for Kohana. Advertising for a symfony dev might get more responses, but no guarantee they'll have the skills / experience you need. If they do, then it really shouldn't be too hard for them to get to grips with Kohana.

If you are going to invest lots of time in rewriting your existing system I would really strongly encourage spending that time making it so it doesn't really matter what framework / database layer / etc it's using. Then you can stick with Kohana, move to symfony, or move to individual components as you wish. And you won't need to be looking for "Kohana devs" or "symfony devs" in future, just php devs. 

Spending effort rewriting to be coupled to a different framework is just wasted work IMO.

**Comment by neo22s** [comment]
@sigmatec I will not be worried about getting decent developers on X framework, any good developer will get used in no time to any of them... my 5 cents.

@shadowhand @acoulton I dont mind if you add me too, maybe is even better if theres 2 different people, but I will not keep it updated ;)

regards

**Comment by shadowhand** [comment]
@neo22s @acoulton you now have Owner role for the entire Kohana org. I have hidden my membership and will no longer actively manage Kohana org unless requested by one of you.

@neo22s feel free to make koseven part of the Kohana org if you wish.

**Comment by shadowhand** [comment]
I was going to close this, but it should really be left open for people to find until the retirement date.

**Comment by neo22s** [comment]
@shadowhand let me talk to the community, let's see what they want. Thanks

**Comment by JackEllis** [comment]
@neo22s From a business point of view, there are situations where companies have their systems in Kohana but can't find developers through the usual channels (e.g. LinkedIn). I think we should consider building a 'Kohana Developers' list that includes contractors / agencies that 'actively' support Kohana. Because the reality is that any of us can jump into a Kohana build and build new business functionality subject to availability. So when companies search 'Kohana Framework developers', there's less struggle. Also, I do agree with the concept of developers getting used to frameworks but a lot of smaller companies that are recruiting developers don't understand that.

@neo22s I'm sure we could host the website on a DO instance or something...

**Comment by neo22s** [comment]
@JackEllis agree! I think a directory is a great idea!

DO to host the site? we use github pages ;)

**Comment by JackEllis** [comment]
@neo22s I was thinking of a "Upload Your Details" php script but I agree it would be better to do it through Github pages!

**Comment by sigma-technology** [comment]
Just in case anyone is losing sleep over not knowing what decision I have made _sarcasm_, after much deliberation and almost being swung to Laravel, I have settled on Symfony.

Yes, I agree it is verbose by comparison, but I find it more intuitive because of the way it is written in almost plain English.  Laravel's `@foreach` etc. syntax in Blade makes me feel rather uncomfortable for some reason as well.

It turns out Data Mapper ORM is actually much more suited to our use-case than Active Record, so Doctrine is ideal for us really due to our database tables not directly representing the objects that get used by the controllers (intentionally; not through bad design).

Thank you to everyone for your feedback, comments and suggestions; they are much appreciated.  Also @shadowhand I would like to echo the other comments on here in terms of a thank you for your work on Kohana in the past, and thanks to @neo22s for keeping it alive as Koseven.

@acoulton I agree it's difficult to find competent PHP devs in the UK.  There are so many courses to learn M$ languages, but hardly any for open source stuff.  I actually started my life as a ASP 3.0 dev, and very quickly defected to PHP 4 when ASP.net came out, which was possibly the worst thing ever created in the history of computing.

**Comment by neo22s** [comment]
@sigmatec 
> quickly defected to PHP 4 when ASP.net came out, which was possibly the worst thing ever created in the history of computing.

oh damn I had exactly the same... first versions of asp.net took sooo many nights of my sleep.... :(

**Comment by acoulton** [comment]
@shadowhand thank you (sorry, was away without internet last week).

**Comment by jaykravetz** [comment]
First of all, A huge thank you to the maintainers and contributors of the Kohana project. Kohana has served us quite well over the last several years and was my own official introduction to PHP frameworks as a new dev.

Hats off to you guys
/standing ovation

I'm wondering if anyone can specify whether or not the github organization and repositories will be taken offline on 07/01 or not

In any event, would it be OK to copy the core & module repositories over to another organization for our own use? (it'll be public so others can use it too if they want)

Thank you

**Comment by shadowhand** [comment]
@jaykravetz the Github organization and repositories will remain indefinitely.

**Comment by jaykravetz** [comment]
@shadowhand thank you for the reply!

What will be going offline?

**Comment by shadowhand** [comment]
@jaykravetz everything not Github.

**Comment by yurii-github** [comment]
i've just tried it on some test things and i would say it is outdated, but 3.4/dev is ok if you replace your ORM with Doctrine and put some other proven components. So yes, it's easy and simple for simple stuff,
If it is just possible to polish 3.4 a bit and release it some people may find it useful.

plz keep repo as it is good reminder how code was evolving for the last decade.

regards
